### PR TITLE
♻️ Is2392/pre iterators: models factory and some pre-iterator changes

### PR DESCRIPTION
**Pull request by pcrespov**
## What do these changes do?

These are mostly to chunks of changes that have been separated from the PR #2556 to simplify the review.

### ⚗️ Preview of new tooling to create models

Maintaining large models representing resource can be challenging, specially when every interface needs a  slightly different variation of the original domain model (e.g. fields are appended/excluded, constraints changes as r/w or r-only, optional/required, nullable, etc). This PR introduces some functional-based tools to overcome these challenges:

 - ``models_library.utils.pydantic_models_factory`` assists the creation of pydantic models, specially those excluding fields (since including/modifying existing fields can be achieved by inheritance). In the associated ``packages/models-library/tests/test_utils_pydantic_models_factory.py`` we demonstrate these tools by building models for ``User`` resource.

In addition, as a first real use case, we introduce some steps to apply the methods above to the ``Project`` resource. As a first step, a pydantic model is automatically generated (see new recipe in ``Makefile``) out of the (language agnostic) json-schema specs: 

 - ``projects._project_models_rest.py`` a  pydantic one-to-one implementation of ``api/specs/common/schemas/project-v0.0.1.json`` (see new ``web/server/Makefile`` recipe). It will be used (soon) in conjunction with ``models_library.utils.models_factory`` utils to implement new variants of the ``Project`` resource model.

### ♻️ ``webserver`` cleanup 
-  refactors ``version_control`` app add-on
-  minor cleanup on other ``webserver``'s addons
   -  ``director_v2_handlers`` 
   -  ``computation_subscribe``
   - ``projects`` cleanup 
   
## Related issue/s

- changes separated from PR #2556  to make the later more focused on the functionality added by the meta plugin

## How to test

```
cd  packages/models-library
make install-dev
pytest -vv tests/test_utils_models_factory.py
```

## Checklist

- [x] Openapi changes? ``make openapi-specs``, ``git commit ...`` and then ``make version-*``)
- [x] Database migration script? ``cd packages/postgres-database``, ``make setup-commit``, ``sc-pg review -m "my changes"``
- [x] Unit tests for the changes exist
- [x] Runs in the swarm
- [x] Documentation reflects the changes
- [x] New module? Add your github username to [.github/CODEOWNERS](.github/CODEOWNERS)

### CODE REVIEW

File: packages/models-library/src/models_library/utils/models_factory.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+#     exclude: Optional[Set[str]] = None
+# ):
+#     selection = _eval_selection(model_cls.__fields__.keys(), include, exclude)
+
+#     # TODO: trim all fields based on selection!
+#     namespace = {}
+#     namespace["fields"] =
+#     namespace["schema_extra"] =
+
+#     return inherit_config(model_cls.__config__, model_cls.__config__, **namespace)
```
Comment (sanderegg) [inline packages/models-library/src/models_library/utils/models_factory.py:127]: are these commented lines necessary?

File: packages/models-library/tests/test_utils_models_factory.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+    username: str
+
+    # parses from User (i.e. validated domain model)
+
+
+class UserListItem(BaseModel):
+    """<- out Item model for response in GET /users"""
+
+    id: PositiveInt
+    username: str
```
Comment (sanderegg) [inline packages/models-library/tests/test_utils_models_factory.py:163]: not sure I understand here how or why you remove display_name. but this is just an example, it's flexible right?

File: services/web/server/src/simcore_service_webserver/computation_subscribe.py [hunk 3]
```diff
@@ -54,6 +54,7 @@ async def progress_message_parser(app: web.Application, data: bytes) -> None:
                 {
                     "event_type": SOCKET_IO_NODE_UPDATED_EVENT,
                     "data": {
+                        "project_id": project["uuid"],
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/computation_subscribe.py:57]: not too sure why this change is here, but I guess it does not hurt??

File: services/web/server/src/simcore_service_webserver/projects/_project_models_rest.py [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+
+class Position1(BaseModel):
+    class Config:
+        extra = Extra.forbid
+
+    x: int = Field(..., description="The x position", example=["12"])
+    y: int = Field(..., description="The y position", example=["15"])
+
+
+class Workbench1(BaseModel):
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/projects/_project_models_rest.py:211]: ok, so if I get it right, these models are auto-generated with your new tool right?
are they in use?
and these suffixes? what do they mean?

File: services/web/server/src/simcore_service_webserver/projects/project_models.py [hunk 5]
```diff
@@ -1,10 +1,12 @@
 from enum import Enum
-from typing import Any, Dict, Optional
+from typing import Any, Dict, Optional, Tuple
 
+from aiopg.sa.result import ResultProxy
 from simcore_postgres_database.models.projects import ProjectType
 
 # TODO: extend
 ProjectDict = Dict[str, Any]
+ProjectProxy = ResultProxy
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/projects/project_models.py:9]: this one, I'm not too sure whether it's good. I mean a ResultProxy is something coming from aiopg and probably we should have converted to a dict instead of passing it around (at least that is my impression). My point is that I'm not sure passing ProjectProxy is very clear.
Whatever... it's just a thought.

File: services/web/server/src/simcore_service_webserver/projects/projects_handlers.py [hunk 6]
```diff
@@ -354,14 +355,20 @@ async def replace_project(request: web.Request):
...
+        ) from err
+
     except KeyError as err:
         raise web.HTTPBadRequest(reason=f"Invalid request parameter {err}") from err
     except json.JSONDecodeError as exc:
         raise web.HTTPBadRequest(reason="Invalid request body") from exc
 
-    # Prune state field (just in case)
-    new_project.pop("state", None)
-
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/projects/projects_handlers.py:364]: ;)

File: services/web/server/src/simcore_service_webserver/version_control_db.py [hunk 7]
```diff
@@ -129,49 +125,83 @@ async def _get_head_branch(
             )
             return branch
 
-    async def _get_head_commit(
+    async def _get_HEAD_commit(
         self, repo_id: int, conn: SAConnection
-    ) -> Optional[RowProxy]:
+    ) -> Optional[CommitProxy]:
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/version_control_db.py:130]: actually same as with ProjectProxy... I wonder. here ideally it should be some domain model going around right? not a aiopg specific object. but that is a minor.

File: services/web/server/src/simcore_service_webserver/version_control_db.py [hunk 8]
```diff
@@ -129,49 +125,83 @@ async def _get_head_branch(
...
     @staticmethod
     async def _upsert_snapshot(
         project_checksum: str,
         project: Union[RowProxy, SimpleNamespace],
         conn: SAConnection,
     ):
+        import json
+
+        from servicelib.json_serialization import json_dumps
+
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/version_control_db.py:195]: import inside the function?

File: services/web/server/src/simcore_service_webserver/version_control_db.py [hunk 9]
```diff
@@ -490,3 +526,46 @@ async def get_snapshot_content(self, repo_id: int, commit_id: int) -> Dict:
                     return snapshot.content
 
         raise NotFoundError(name="snapshot for commit", value=(repo_id, commit_id))
+
+    async def get_workbench_view(self, repo_id: int, commit_id: int) -> Dict[str, Any]:
+        # FIXME: q&d fix!!!
+        # TODO: move to VersionControlRepositoryInternalAPI
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/version_control_db.py:532]: I am not too sure about this kind of comments... are you sure it's ready?

File: services/web/server/src/simcore_service_webserver/version_control_db.py [hunk 10]
```diff
@@ -490,3 +526,46 @@ async def get_snapshot_content(self, repo_id: int, commit_id: int) -> Dict:
...
+                )
+                assert repo  # nosec
+
+                # if snapshot differs from wcopy, then show working copy
+                wcopy_project_id = await self._fetch_wcopy_project_id(
+                    repo_id, commit_id, conn
+                )
+
+                # FIXME: in general this is wrong. For the moment,
+                # all wcopies except for the repo's main wcopy (i.e. repo.project_uuid) are READ-ONLY
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/version_control_db.py:552]: same here... also I don't know if I understand them...

File: services/web/server/tests/unit/with_dbs/docker-compose-devel.yml [hunk 11]
```diff
@@ -1,7 +1,8 @@
 version: "3.8"
 services:
   postgres:
-    image: "postgres:10.11@sha256:2aef165ab4f30fbb109e88959271d8b57489790ea13a77d27c02d8adb8feb20f"
+    image: "postgres:10.11@sha256:2aef165ab4f30fbb109e88959271d8b57489790ea13a77d27\
+      c02d8adb8feb20f"
```
Comment (sanderegg) [inline services/web/server/tests/unit/with_dbs/docker-compose-devel.yml:5]: this is a bit weird... is that autoformatting?

File: services/web/server/tests/unit/with_dbs/docker-compose-devel.yml [hunk 12]
```diff
@@ -20,16 +21,23 @@ services:
     #   - net.ipv4.tcp_keepalive_intvl=600
     #   - net.ipv4.tcp_keepalive_probes=9
     #   - net.ipv4.tcp_keepalive_time=600
-    command: postgres -c tcp_keepalives_idle=600 -c tcp_keepalives_interval=600 -c tcp_keepalives_count=5
+    command: postgres -c tcp_keepalives_idle=600 -c tcp_keepalives_interval=600 -c
+      tcp_keepalives_count=5
```
Comment (sanderegg) [inline services/web/server/tests/unit/with_dbs/docker-compose-devel.yml:25]: what about:
```suggestion
    command: |
      postgres 
      -c tcp_keepalives_idle=600 
      -c tcp_keepalives_interval=600 
      -c tcp_keepalives_count=5
```

**Review by sanderegg — APPROVED** [review]
a few comments, but this is looking promising!

File: services/web/server/src/simcore_service_webserver/computation_subscribe.py (hunk 3 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/computation_subscribe.py:57]: this is part of the preparations for the iterator (therefore pre-iterator in the title) 

When running a meta-project, it will run tasks from different iteration projects simultaneously. Therefore a new situation arises in which the active project is not the only one that runs and if the front-end wants to compute the progress of the meta-projects it needs to add up each progress of the iteration projects

File: services/web/server/src/simcore_service_webserver/projects/project_models.py (hunk 5 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/projects/project_models.py:9]: @sanderegg actually it is a mistake. It is a ``RowProxy`` and not ``ResultProxy``

I need to do all these aliases because right now it is very difficult to figure out what "type" of datastructure is holding the ``Project``. 

NOTE: aiopg does not covert it to dict ... see some trials i did in [test_utils_aiopg_orm](https://github.com/ITISFoundation/osparc-simcore/blob/master/packages/postgres-database/tests/test_utils_aiopg_orm.py#L278)

File: services/web/server/tests/unit/with_dbs/docker-compose-devel.yml (hunk 11 above)
Comment (pcrespov) [reply services/web/server/tests/unit/with_dbs/docker-compose-devel.yml:5]: yes :-/

File: packages/models-library/tests/test_utils_models_factory.py (hunk 2 above)
Comment (pcrespov) [reply packages/models-library/tests/test_utils_models_factory.py:163]: it is an example. I just wanted to create a "lighter"/"non-detailed" version of ``UserGet`` for the lists to demostrate the flexibility

File: services/web/server/src/simcore_service_webserver/projects/_project_models_rest.py (hunk 4 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/projects/_project_models_rest.py:211]: no, no ... :-(  I guess i did not explain it clearly in the description

This is created with [``datamodel-codegen``](https://github.com/koxudaxi/datamodel-code-generator)  from json-schema. See services/web/server/Makefile.

```
simcore_service_webserver/api/v0/schemas/project-v0.0.1.json --(datamodel-codegen)-->_project_models_rest.py
```
then this module is used as ``reference_cls`` to build all model variants of the ``Project`` resource. And it is here where we use ``models_factory.copy`` to build the model classes.

@sanderegg it is important that this idea gets through and we agree on it.

File: services/web/server/src/simcore_service_webserver/version_control_db.py (hunk 8 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/version_control_db.py:195]: good catch!

File: services/web/server/src/simcore_service_webserver/version_control_db.py (hunk 9 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/version_control_db.py:532]: I forgot to cleanup some of those

File: services/web/server/src/simcore_service_webserver/version_control_db.py (hunk 10 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/version_control_db.py:552]: These are all preparations and shall be resolved in next PR, i.e. #2556

File: services/web/server/src/simcore_service_webserver/version_control_db.py (hunk 7 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/version_control_db.py:130]: In this case, I disagree:
1. this is a protected function used as a wrapper around ``aiopg.sa`` interface, therefore outputs are ``*Proxy``-like data structures
2. This is an alias of ``RowProxy`` to clarify the table we are getting
3. i/o public methods of ``VersionControlRepository`` do have their own models (e.g. CommitLog) and yes, those know nothing about ``aiopg.sa`` models

File: packages/models-library/src/models_library/utils/models_factory.py (hunk 1 above)
Comment (pcrespov) [reply packages/models-library/src/models_library/utils/models_factory.py:127]: Oh boy, I was not very careful. I will do another check to make sure nothing more gets in. thx!

File: services/web/server/src/simcore_service_webserver/projects/_project_models_rest.py (hunk 4 above)
Comment (mguidon) [reply services/web/server/src/simcore_service_webserver/projects/_project_models_rest.py:211]: I am also not 100 % sure I understand. So you suggest to parse all json-schemas that we have with the code-generator and that would give the baseline models (100% complete with all fields). Those will replace the various models.py in services and model library? And then from there you create specialized model classes (omitting for instance some fields). Is this done on the fly or are you running your tool over them and store the results in a static file? What happens if you want to add fields wrt to the baseline model?

Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/projects/_project_models_rest.py:211]: @mguidon here some clarifications

> I am also not 100 % sure I understand. So you suggest to parse all json-schemas that we have with the code-generator and that would give the baseline models (100% complete with all fields). Those will replace the various models.py in services and model library? 

Basically yes. This aligns as well with "specifications driven design". We use as "source of truth" the language-agnostic json-schemas (or OAS) specs . Then we "translate" them into pydantic models by auto-generating them into reference models (this is how i denote what you call "baseline models"). BTW we could actually do this also at runtime instead of hard-coding it.

> And then from there you create specialized model classes (omitting for instance some fields). Is this done on the fly or are you running your tool over them and store the results in a static file? 

This is done at runtime (IMO this does not has an overhead) but it could be relatively easy to extend it to produce "hard-coded" classes. This would definitively make more clear "the final model class" .  The only drawback with that approach is that we need to re-generate and commit every time there is a change and my experience tells me that we are not very good at that :-) 

> What happens if you want to add fields wrt to the baseline model?

You can use the generated class as a base class as well (see usage examples in ``packages/models-library/tests/test_utils_models_factory.py``)

File: packages/models-library/src/models_library/utils/pydantic_models_factory.py [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+#     exclude: Optional[Set[str]] = None
+# ):
+#     selection = _eval_selection(model_cls.__fields__.keys(), include, exclude)
+
+#     # TODO: trim all fields based on selection!
+#     namespace = {}
+#     namespace["fields"] =
+#     namespace["schema_extra"] =
+
+#     return inherit_config(model_cls.__config__, model_cls.__config__, **namespace)
```
Comment (GitHK) [inline packages/models-library/src/models_library/utils/pydantic_models_factory.py:144]: will this be used?

File: packages/models-library/tests/test_utils_pydantic_models_factory.py [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+
+
+def test_build_UserReplace_model():
+    # model for request body Replace method https://google.aip.dev/134
+
+    # Replace is like create but w/o primary key (if it would be defined on the client)
+    class _UserReplace(copy_model(User, exclude={"id", "password_hash"})):
+        password: str
+        password2: str
+        _passwords_match = create_validator_for("password2")
```
Comment (GitHK) [inline packages/models-library/tests/test_utils_pydantic_models_factory.py:240]: I'm commenting here to not loose track of it.

The following will be great to have, for development reasons:
Please try to have some sort of utility functions which allows me to discover all the related models which have been copied. and their source. Please keep in mind to support the copy of a copy.

File: scripts/json-schema-dereference.bash [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+set -o pipefail
+IFS=$'\n\t'
+
+image_name="$(basename $0):latest"
+
+# derefernce json-schemas for easy comparison
+# SEE https://github.com/davidkelley/json-dereference-cli
+docker build --tag "$image_name" - <<EOF
+FROM node:12.18.2
+RUN npm install -g json-dereference-cli
```
Comment (GitHK) [inline scripts/json-schema-dereference.bash:14]: maybe having the version pinned of this `json-deference-cli` would help in the future.

**Review by GitHK — APPROVED** [review]
Pair reviewed. Just a few things to check.

Note: If you did not plan to add it now, maybe create an issue for a function to auto discover copied models. Just to not loose track of it.

File: packages/models-library/src/models_library/utils/pydantic_models_factory.py (hunk 13 above)
Comment (pcrespov) [reply packages/models-library/src/models_library/utils/pydantic_models_factory.py:144]: no

File: packages/models-library/tests/test_utils_pydantic_models_factory.py (hunk 14 above)
Comment (pcrespov) [reply packages/models-library/tests/test_utils_pydantic_models_factory.py:240]: added in open issue together with other ideas
